### ISSUE TITLE
Make collectionMappingStrategy globally configurable

### ISSUE THREAD

**Issue by joshiste** [mapstruct/mapstruct#1961]
I'd love to configure the collectionMappingStrategy for all Mappers via setting a annotation processor option

**Comment by Captain1653** [comment]
Hi, @joshiste . Thanks for your ticket. We discussed it before in https://github.com/mapstruct/mapstruct/issues/263  . Is it really necessary this annotation processor option for your?

**Comment by joshiste** [comment]
>Can we use some common @MapperConfig ?

IIRC i'd then have to refer to that MapperConfig in each and every of my mappers or is there a way to set a global MapperConfig for all my mappers globally?

**Comment by Captain1653** [comment]
With @MapperConfig I think yes. With new implementation - I'm not sure (It depends on the implementation: @GlobalConfig or @MapperConfig.isGlobal). Honestly, (my opinion) I like to use a processor option. It's more clearly for me. @sjaakd  @filiphr what do you think? Should we do this option? I we need it I can do it.

**Comment by sjaakd** [comment]
I'm not in favour of adding options to the processor directly. I prefer a more programatic approach. However, I do like the idea of having a global config. Perhaps that's something you can point out as an option. 

The question is what we should do when somebody overrides this in a `Mapper`. Accept the one there? Error?

Anyway: @filiphr WDYT?

**Comment by Captain1653** [comment]
With overrides can be some conflict (maybe I'm wrong, correct me please). Our common case: mapper takes precedence on @MapperConfig. Will it  be inverted with @GlobalConfig? Or not?

**Comment by sjaakd** [comment]
Well.. I would not propose to have another config. So just stick to `@MapperConfig`. But the idea is that you can specify it in the annotation processing definition

Like this:

```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-compiler-plugin</artifactId>
    <version>3.5.1</version>
    <configuration>
        <source>1.8</source>
        <target>1.8</target>
        <annotationProcessorPaths>
            <path>
                <groupId>org.mapstruct</groupId>
                <artifactId>mapstruct-processor</artifactId>
                <version>${org.mapstruct.version}</version>
            </path>
        </annotationProcessorPaths>
        <compilerArgs>
            <compilerArg>
                -Amapstruct.globalconfigurationclass=com.example.MyConfig
            </compilerArg>
        </compilerArgs>
    </configuration>
</plugin>
```

```java
@MapperConfig(unmappedTargetPolicy = ReportingPolicy.WARN)
public interface MyConfig{
// config stuff
}
```

All mappers would then take this config instead of explicitly having to specify it. 

We just have to define what happens when a mapper does specify a config. Override is the best option I guess..

**Comment by Captain1653** [comment]
Thank you, for example, @sjaakd  . I think, you are right. Override is the good idea. You know, we have another case (I suppose). What about the conflict between -Amapstruct.globalConfigurationClass and others configuration options? For example -Amapstruct.unmappedTargetPolicy. Will be result from global config or from the configuration property?

**Comment by sjaakd** [comment]
> hat about the conflict between `-Amapstruct.globalConfigurationClass` and others configuration options? For example `-Amapstruct.unmappedTargetPolicy`, like in your example.

Ahh. would that be any different than is the case now? I think they are overridden (check?) by any programmatic definition. So, the same would happen if you define a global config here. Of course, we need to warn for that.

**Comment by filiphr** [comment]
I would not go with a `globalConfigurationClass`. Since we can't really load such a class dynamically during the compilation. I would say that if we add something we add it as a compiler property like `mapstruct.collectionMappingPolicy`. 

Having said all of this, perhaps it is an interesting idea to have an SPI for providing the default properties for MapStruct. There is something similar in AssertJ, have a look at [`ConfigurationProvider`](https://github.com/joel-costigliola/assertj-core/blob/master/src/main/java/org/assertj/core/configuration/ConfigurationProvider.java) and [`Configuration`](https://github.com/joel-costigliola/assertj-core/blob/master/src/main/java/org/assertj/core/configuration/Configuration.java).

And of course if something is programatically defined via `@MappingConfig`, `@Mapper`, `@BeanMapping`, etc. it would have preference.

**Comment by Captain1653** [comment]
So, @sjaakd and @filiphr . If you have the same opinion I can implement it. It's not so difficult. Should I do it?

**Comment by sjaakd** [comment]
> Since we can't really load such a class dynamically during the compilation. I would say that if we add something we add it as a compiler property like `mapstruct.collectionMappingPolicy`.

why not? Check out: `MethodRetrievalProcessor`.. We could do a `elementUtils#getTypeElement`... right?

**Comment by filiphr** [comment]
Yes you are right @sjaakd. It could work like that.

However, to me it doesn't feel the right approach to do it. I think that having an SPI is the way to go for this. Since referencing a class via a property is more or less similar to this. The SPI can also be in the `mapstruct` module and not the `mapstruct-processor` module. 

In my opinion having an SPI could allow people to define their won dependency and use it company wide without the need to define anything in their properties. Just the service being there would be enough.

**Comment by sjaakd** [comment]
>  think that having an SPI is the way to go for this.

Hmm. I'm not convinced. I think we should device a strategy when to use SPI. As you know, it was intended to write implementations behind a standard API. I guess MapStruct takes this a bit further currently. Having to put your SPI in a seperate module (my experience?) does not help either.

Couldn't we say that SPI is used for organization wide strategy (e.g. an organization often chooses to go for lombok, have different conventions, etc. 
And use configuration more for a project / module wide strategy?

**Comment by sjaakd** [comment]
> think that having an SPI is the way to go for this.

@Captain1653 .. Are you working on this one?

**Comment by Captain1653** [comment]
@sjaakd  Yes, I'm working with it. It's little hard for me, but I'm really want to try. I'll show my draft soon.

**Comment by wgottschalk** [comment]
Hi there! Are there any updates on this? My team is adopting the library and what we want to do is to globally configure the
`implementationPackage` so we don't have to include the customConfig in every @Mapping annotation.

If there's an alternative to this I would love to hear about it. Thank you!

**Comment by Kamui123** [comment]
We have excatly the same problem as wgottschalk describe. Any news on this?

**Comment by FireRedDev** [comment]
> Well.. I would not propose to have another config. So just stick to `@MapperConfig`. But the idea is that you can specify it in the annotation processing definition
> 
> Like this:
> 
> ```
> <plugin>
>     <groupId>org.apache.maven.plugins</groupId>
>     <artifactId>maven-compiler-plugin</artifactId>
>     <version>3.5.1</version>
>     <configuration>
>         <source>1.8</source>
>         <target>1.8</target>
>         <annotationProcessorPaths>
>             <path>
>                 <groupId>org.mapstruct</groupId>
>                 <artifactId>mapstruct-processor</artifactId>
>                 <version>${org.mapstruct.version}</version>
>             </path>
>         </annotationProcessorPaths>
>         <compilerArgs>
>             <compilerArg>
>                 -Amapstruct.globalconfigurationclass=com.example.MyConfig
>             </compilerArg>
>         </compilerArgs>
>     </configuration>
> </plugin>
> ```
> 
> ```java
> @MapperConfig(unmappedTargetPolicy = ReportingPolicy.WARN)
> public interface MyConfig{
> // config stuff
> }
> ```
> 
> All mappers would then take this config instead of explicitly having to specify it.
> 
> We just have to define what happens when a mapper does specify a config. Override is the best option I guess..

Is this finished now? I would love to configure the componentModel to spring via @MapperConfig globally

**Comment by rubenmartinez** [comment]
➕

**Comment by filiphr** [comment]
>Is this finished now?

@rubenmartinez this issue is still open, so no this isn't finished. 

>I would love to configure the componentModel to spring via @MapperConfig globally

@rubenmartinez this is already possible using the already available compiler option `mapstruct.defaultComponentModel`. See more in the [Configuration Options](https://mapstruct.org/documentation/stable/reference/html/#configuration-options) section of our documentation.
